rust/connlib/clients/shared/src/lib.rs

```diff
@@ -280,38 +153,62 @@ where
-    interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
-
-    interval
-}
-
-/// Parses an interval from the _compile-time_ env variable `CONNLIB_LOG_UPLOAD_INTERVAL_SECS`.
+/// Connects to the portal and starts a tunnel.
 ///
-/// If not present or parsing as u64 fails, we fall back to a default interval of 5 minutes.
-fn upload_interval_duration_from_env_or_default() -> Duration {
-    const DEFAULT: Duration = Duration::from_secs(60 * 5);
-
-    let Some(interval) = option_env!("CONNLIB_LOG_UPLOAD_INTERVAL_SECS") else {
-        tracing::warn!(interval = ?DEFAULT, "Env variable `CONNLIB_LOG_UPLOAD_INTERVAL_SECS` was not set during compile-time, falling back to default");
-
-        return DEFAULT;
+/// When this function exits, the tunnel failed unrecoverably and you need to call it again.
+async fn connect<CB>(
+    api_url: Url,
+    token: SecretString,
+    device_id: String,
+    device_name_override: Option<String>,
+    os_version_override: Option<String>,
+    callbacks: CB,
+    max_partition_time: Option<Duration>,
+) where
+    CB: Callbacks + 'static,
+{
+    let (portal_url, private_key) = match login_url(
+        Mode::Client,
+        api_url,
+        token,
+        device_id,
+        device_name_override,
+    ) {
+        Ok((portal_url, private_key)) => (portal_url, private_key),
+        Err(e) => {
+            tracing::error!("Failed to build login URL: {e}");
+            let _ = callbacks.on_disconnect(&e);
+            return;
+        }
     };
 
-    let interval = match interval.parse() {
-        Ok(i) => i,
+    let tunnel = match Tunnel::new(private_key, callbacks.clone()) {
+        Ok(tunnel) => tunnel,
         Err(e) => {
-            tracing::warn!(interval = ?DEFAULT, "Failed to parse `CONNLIB_LOG_UPLOAD_INTERVAL_SECS` as u64: {e}");
-            return DEFAULT;
+            tracing::error!("Failed to make tunnel: {e}");
+            let _ = callbacks.on_disconnect(&e);
+            return;
         }
     };
 
-    tracing::info!(
-        ?interval,
-        "Using upload interval specified at compile-time via `CONNLIB_LOG_UPLOAD_INTERVAL_SECS`"
+    let portal = PhoenixChannel::connect(
+        Secret::new(SecureUrl::from_url(portal_url)),
+        get_user_agent(os_version_override),
+        PHOENIX_TOPIC,
+        (),
+        ExponentialBackoffBuilder::default()
+            .with_max_elapsed_time(max_partition_time)
+            .with_max_interval(MAX_RECONNECT_INTERVAL)
+            .build(),
     );
 
-    Duration::from_secs(interval)
+    let mut eventloop = Eventloop::new(tunnel, portal);
+
+    match std::future::poll_fn(|cx| eventloop.poll(cx)).await {
+        Ok(never) => match never {},
+        Err(e) => {
+            tracing::error!("Eventloop failed: {e}");
+            let _ = callbacks.on_disconnect(&Error::PortalConnectionFailed); // TMP Error until we have a narrower API for `onDisconnect`
```

Comment:
The eventloop only ever returns fatal errors from the phoenix-channel. Those are being refactored in https://github.com/firezone/firezone/pull/4015 to really only be fatal errors.